### ISSUE TITLE
[BUG] AnimatePresence height animation with padding doesn't work

### ISSUE THREAD

**Issue by snowl** [framer/motion#368]
**1. Read the [FAQs](#faqs) 👇**

**2. Describe the bug**

When using AnimatePresence with some padding in the div that you are animating w/ height animation, the height calculation breaks and only animates the height to the (internal height - padding), both in and out.

**3. IMPORTANT: Provide a CodeSandbox reproduction of the bug**

https://codesandbox.io/s/framer-motion-animatepresence-padding-issue-zqxrk?fontsize=14

If you remove the padding: within the div css, you can see how much more smooth the animation is.

**4. Steps to reproduce**

Animate height on a div that has padding.

**5. Expected behavior**

Height should go from 0 to the height calculated by the browser for "auto".

**6. Video or screenshots**

Here['s](https://gfycat.com/redneedybandicoot) a gif of how it looks in my application: https://gfycat.com/weirdachingdogwoodclubgall

You can see the div underneath "snaps" because of all the height that still exists due to the padding.

**7. Environment details**

Tested in chrome

**Comment by thebuilder** [comment]
I don't think this a framer motion bug. You'd see this regardless of how you are animating height.
It's how the browsers deal with padding when height is 0.

**Comment by snowl** [comment]
I don't think so - if you slow down the transitions down so you can observe the values they end up animating from "auto" to the total padding height instead of animating from "auto" to 0. Also - even if you try and set the animation to:

```
<motion.div
                key={item}
                animate={{ height: "auto", opacity: 1, paddingTop: 30, paddingBottom: 30 }}
                exit={{ height: 0, opacity: 0, paddingTop: 0, paddingBottom: 0 }}
                initial={{ height: 0, opacity: 0, paddingTop: 0, paddingBottom: 0 }}
              >
```

Not only does it exhibit the same behavior, but it also now "snaps" in when animating in as well. I'm fairly sure that this should be animating smoothly here.

**Comment by Rijk** [comment]
I have noticed this behaviour as well when testing with `height: 'auto'`. I also think it’s a Framer Motion measuring bug.

**Comment by vavra7** [comment]
Any updates on this? The problem is still present.

**Comment by cjoecker** [comment]
I used until this is fixed a work around with Material-UI transitions:

```
import { Collapse } from '@material-ui/core';
import { motion, useIsPresent } from 'framer-motion';
import React, { memo, useEffect, useState } from 'react';

interface CollapseAnimationProps {
  id: number;
  children: React.ReactNode;
}

function _CollapseAnimation({ id, children }: CollapseAnimationProps) {
  const isPresent = useIsPresent();
  const [isVisible, setIsVisible] = useState(false);

  const ANIMATION_TIME = 500;

  useEffect(() => {
    const timeout = setTimeout(() => setIsVisible(true), 0);
    return () => clearTimeout(timeout);
  }, []);

  useEffect(() => {
    if (isVisible && !isPresent) {
      setIsVisible(false);
    }
  }, [isPresent, isVisible]);

  return (
    <motion.div
      key={id}
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={{ opacity: 0 }}
      transition={{ duration: ANIMATION_TIME / 1000 }}
    >
      <Collapse in={isVisible} timeout={ANIMATION_TIME}>
        {children}
      </Collapse>
    </motion.div>
  );
}
//Don't use framer motion height animation until this bug is solved:
//https://github.com/framer/motion/issues/368

export const CollapseAnimation = memo(_CollapseAnimation);
```

**Comment by ChrisBrownie55** [comment]
Can confirm this is not the browser's fault. I slowed down the animation and it looks like Framer Motion is stopping the animation for the height at an early number. Even if you animate the padding to `0` it stops early. For me it's stopping at like `20px` for no reason. If I take away all padding animations it works fine.

**Comment by mattgperry** [comment]
Yeah it's definitely a mis-measurement because of the padding. I'll have to look into it.

**Comment by ardiewen** [comment]
For now, this can be solved with `height: auto` animating on an outside container without padding, and set padding on an inner container.

**Comment by MikeWillis** [comment]
For me @ardiewen's solution worked when collapsing a div, but unfortunately not when opening it. Still better though

**Comment by oORiggsOo** [comment]
I'm having the same issue, except it seems to fail on the first <motion.div animate={{}}...  but works when moved to the next <motion.h2 animate={{}}......      My browser spits this warning out when it fails,,,,    

`"You are trying to animate marginTop from "100px" to "100". 100px is not an animatable value - to enable this animation set 100px to a value animatable to 100 via the `style` property."`

Here's what causes the fail. First <motion>

               ```
                  <motion.div
			animate={{ opacity: 0.2, x: "100px", y: "100", marginTop: "100" }}
		  >
                         // all code from here on works, 
                         <motion.h2 ...animate={{code.........
                        ..............................................................................
                </motion.div>
               ```

**Comment by kylemh** [comment]
This bug extends to border as well.

**Comment by mattgperry** [comment]
I'm just looking into this now. The behaviour in the original post isn't a bug. When `height` is animated to `0`, *height* is animated to `0`. But of course with CSS that doesn't mean the actual height is going to be `0` because padding still exists.

The snapping in this example: https://github.com/framer/motion/issues/368#issuecomment-544305139 is a bug and maybe something we can do something about

**Comment by mattgperry** [comment]
https://github.com/framer/motion/pull/1321#issuecomment-964199551

**Comment by present-g** [comment]
Hi! When can we expect a fix?

**Comment by amogadasi** [comment]
moving the padding to an inner div solved it for me:

https://codesandbox.io/s/framer-motion-animatepresence-padding-issue-forked-317ykj?

**Comment by marcospassos** [comment]
This bug makes the DX really bad. Using borders, margins, and padding is almost unavoidable for use cases like lists. Currently require a lot of boilerplate code to make it work smoothly.

**Comment by soufDev** [comment]
hello 👋 @mattgperry 

I tried your workaround, unfortunately the animation is not really working if we add padding to the inner div 😢 

I have to delete it to have the smooth animation

You can have a look at the video just bellow

Please I would really need help on this one 🙏 

https://user-images.githubusercontent.com/15775944/183764478-889cb094-f42d-47bf-8a01-05f6bee97943.mp4

**Comment by minlare** [comment]
I had this issue when using `display: grid` and `row-gap`. Tweaking the html to add a wrapping element to each list item with `padding` instead of `row-gap` works perfectly!

**Comment by adomaitisc** [comment]
I was having the same issue of jumping animation with padding.
Then, I decided to not use padding at all.

I am using empty p tags with the height of the padding I need.

<img width="560" alt="image" src="https://user-images.githubusercontent.com/57054660/210947315-b3d2c637-3a8e-4062-9ec5-6abc40fee1a2.png">

I can say that it works.

Not a huge workaround but not pretty as well. And it is only a 'hack' until something official comes in. 

Thanks!

**Comment by merlindru** [comment]
If it still doesn't work, one way to solve this is to use something like @react-hook/size, to measure the size of your **inner div** and then pass that to framer-motion explicitly:

```tsx
const ref = useRef<HTMLDivElement | null>(null);
const [width, height] = useSize(ref);

<motion.section
    animate={active ? "active" : "idle"}
    variants={{
        idle: { height: 50 },
        active: { height: height },
    }}
>
    <motion.div ref={ref}> ... </motion.div>
</motion.section>
```
